instance/models/instance.py

```diff
@@ -140,22 +216,21 @@ def status(self):
         return None
 
     @property
-    def progress(self):
-        """
-        Instance's current status progress
-        """
-        server = self._current_server
-        if server:
-            return server.progress
-        return None
-
-    @property
     def event_context(self):
         """
         Context dictionary to include in events
         """
         return {'instance_id': self.pk}
 
+    def _transition(self, state_transition):
+        """
+        Helper method to update the state of this instance.
+
+        Ensures that state is in sync with the Django DB.
+        """
+        state_transition()
+        self.save()
```

Comment:
Hmm, maybe we should just move this logic to `ModelResourceStateDescriptor`.

You could put it into `_set_state` as follows:

```python
def _set_state(self, resource, new_state_class):
    ...
    # Save changes to this one field only:
    resource.save(update_fields=[self.model_field_name]) 
    return new_state
```

Then you can remove the `_transition` method from Instance and Server, and we can just call the transition methods (like self._status_to_waiting_for_server()) directly.